sw/cheri/tests/pinmux_tests.hh

```diff
@@ -0,0 +1,546 @@
+  uart->init();
+  uart->fifos_clear();
+  uart->parity();
+  reset_mcycle();
+
+  // Wait for the UART TX FIFO to be empty and all bits to be transmitted,
+  // and the RX to be idle.
+  while ((uart->status & UartStatusRegTxIdle) == 0 || (uart->status & UartStatusRegRxIdle) == 0) {
+    asm("");
+  }
+  uart->fifos_clear();
+
+  // Send and receive a test message, character by character
+  for (uint32_t i = 0; i < test_length; ++i) {
+    const uint8_t test_byte = prng();
+    uart->blocking_write(test_byte);
+
+    // Try to read the character back within the given timeout.
+    const uint32_t start_mcycle   = get_mcycle();
+    const uint32_t cycles         = read_timeout_msec * CyclesPerMillisecond;
+    const uint32_t timeout_mcycle = start_mcycle + cycles;
+    while (!uart->can_read() && get_mcycle() < timeout_mcycle) {
+      asm("");
+    }
+    if (!uart->can_read() || uart->blocking_read() != test_byte) {
+      // On timeout or an incorrect read, fail immediately.
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Tests that the given GPIO pins appear to be working as expected, by sending
+ * and receiving some data using the specified output and input GPIO pins. To
+ * pass, this test expects that the specified input and output GPIO pins are
+ * manually connected on the board, so that GPIO output can be written and
+ * read back by the board.
+ *
+ * @param wait_msec The time in milliseconds to wait for the GPIO write signal
+ * to be transmitted before reading.
+ * @param test_length The number of individual write/reads to test on the given
+ * GPIO connection.
+ *
+ * Returns true if the test passed, or false if it failed.
+ */
+static bool gpio_write_read_test(Capability<volatile SonataGpioFull> gpio, GpioPin output_pin, GpioPin input_pin,
+                                 uint32_t wait_msec, uint32_t test_length) {
+  bool gpio_high = false;
+  for (uint32_t i = 0; i < test_length; i++) {
+    // For each test iteration, invert the GPIO signal
+    gpio_high = !gpio_high;
+
+    // Write this inverted signal on the configured output GPIO
+    set_gpio_output(gpio, output_pin, gpio_high);
+
+    // Wait a short time for the transmission
+    const uint32_t cycles     = wait_msec * CyclesPerMillisecond;
+    const uint32_t end_mcycle = get_mcycle() + cycles;
+    while (get_mcycle() < end_mcycle) {
+      asm("");
+    }
+
+    // Read the GPIO value back
+    if (get_gpio_input(gpio, input_pin) != gpio_high) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * Attempt to retrieve the JEDEC ID of the Spi Flash device on the Sonata board,
+ * and compare it to the known (expected) JEDEC ID. This is separate from the
+ * similar test in the SPI tests, as it skips the initialisation logic that
+ * should not be repeated in the pinmux check.
+ *
+ * Returns the number of failures during the test.
+ */
+static int spi_jedec_id_test(Capability<volatile SonataSpi> spi, SpiFlash spi_flash) {
```

Comment:
`spi` doesn't appear to be used in this function.

Reply:
Looks like I somehow lost the logic that (re)configures the SPI along the way, will add that now.